astropy/units/quantity.py

```diff
@@ -407,8 +407,11 @@ def __new__(cls, value, unit=None, dtype=np.inexact, copy=True, order=None,
                 subok=False, ndmin=0):
 
         if unit is not None:
+            UNIT_IS_NONE = False
```

Comment:
I don't love adding in the flag, but there wasn't an obviously cleaner way, because the ``unit`` input argument is not inviolate.